File: apps/android/feature/ai/data/src/main/java/com/scribblefit/feature/ai/data/engine/DynamicLLMEngine.kt
package com.scribblefit.feature.ai.data.engine

import com.scribblefit.core.config.domain.ConfigRepository
import com.scribblefit.core.config.domain.LLMProvider
import com.scribblefit.feature.ai.domain.LLMEngine
import com.scribblefit.feature.ai.domain.ParsedWorkoutResult
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.withContext

internal class DynamicLLMEngine(
    private val openAIEngine: LLMEngine,
    private val geminiEngine: LLMEngine,
    private val localEngine: LocalAIEngine,
    configRepository: ConfigRepository,
    coroutineDispatcher: CoroutineDispatcher
) : LLMEngine,
    CoroutineScope by CoroutineScope(coroutineDispatcher + CoroutineName("DynamicLLMEngine")) {
    private val activeEngine = configRepository.config.map {
        when (it.preferredLlmProvider) {
            LLMProvider.OPENAI -> openAIEngine
            LLMProvider.GEMINI -> geminiEngine
            else -> localEngine
        }
    }.stateIn(this, SharingStarted.Eagerly, localEngine)

    override suspend fun parseWorkout(rawText: String): Result<ParsedWorkoutResult> =
        withContext(coroutineContext) {
            activeEngine.value.parseWorkout(rawText)
        }
}

File: apps/ios/ScribbleFit/Sources/Features/AI/Data/Engines/DynamicLLMEngine.swift
import Foundation

public final class DynamicLLMEngine: LLMEngine, Sendable {
    private let configRepository: any ConfigRepository
    private let geminiEngine: GeminiAIEngine
    private let openAIEngine: OpenAIAIEngine
    private let localEngine: LocalAIEngine
    private let proxyEngine: ScribbleFitProxyEngine

    public init(
        configRepository: any ConfigRepository,
        geminiEngine: GeminiAIEngine,
        openAIEngine: OpenAIAIEngine,
        localEngine: LocalAIEngine,
        proxyEngine: ScribbleFitProxyEngine
    ) {
        self.configRepository = configRepository
        self.geminiEngine = geminiEngine
        self.openAIEngine = openAIEngine
        self.localEngine = localEngine
        self.proxyEngine = proxyEngine
    }

    public func parseWorkout(rawText: String) async -> ParsedWorkoutResult {
        let config = await configRepository.getConfig()
        switch config?.preferredLlmProvider ?? .proxy {
        case .openai: return await openAIEngine.parseWorkout(rawText: rawText)
        case .gemini: return await geminiEngine.parseWorkout(rawText: rawText)
        case .local: return await localEngine.parseWorkout(rawText: rawText)
        case .proxy: return await proxyEngine.parseWorkout(rawText: rawText)
        }
    }
}
